sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -509,6 +509,10 @@ case class Cast(
     evalMode == EvalMode.TRY
   }
 
+  private def isUserSpecified: Boolean = {
+    getTagValue(Cast.USER_SPECIFIED_CAST).getOrElse(false)
```

Comment:
We shouldn't rely on tree node tag to define the behavior, as tree node tag is unstable today, and we only use it for non-important use cases, such as pretty print and error message.